trecs/models/mf.py

```diff
@@ -0,0 +1,293 @@
+        representations will be assigned randomly.
+
+        >>> mf = ImplicitMF()
+        >>> mf.users_hat.shape
+        (100, 10)   # <-- 100 users (default), 10 latent features (default)
+        >>> mf.items.shape
+        (10, 1250) # <-- 10 latent features (default), 1250 items (default)
+
+        This class can be customized either by defining the number of users/items/latent
+        features in the system.
+
+        >>> mf = ImplicitMF(num_users=1200, num_items=5000)
+        >>> mf.users_hat.shape
+        (1200, 10) # <-- 1200 users, 10 attributes
+
+        >>> mf = ImplicitMF(num_users=1200, num_items=5000, num_latent_features=2000)
+        >>> mf.users_hat.shape
+        (1200, 2000) # <-- 1200 users, 2000 attributes
+
+        Note that all arguments passed in at initialization must be consistent -
+        otherwise, an error is thrown. For example, one cannot pass in
+        `num_users=200` but have `user_representation.shape` be `(300, 100)`.
+        Likewise, one cannot pass in `num_items=1000` but have
+        `item_representation.shape` be `(100, 500)`.
+
+        To use the model, it's advised to use the `startup_and_train` method, which
+        will collect user interaction data from random recommendations. At the end
+        of the startup period, an implicit matrix factorization model is fit to the
+        interaction data, generating new user/item latent representations. These
+        representations are fixed and used for future calls to `run`.
+
+        >>> mf = ImplicitMF(num_latent_features=10) # 10 latent features
+        >>> mf.startup_and_train(50) # 50 timesteps of training
+        >>> mf.run(100)
+
+        If you'd like to retrain the matrix factorization model at any point, you
+        may use the `fit_mf` method. It will fit a new set of latent user/item
+        representations based on the interaction data from the *most recent* call to
+        `run()`.
+
+        >>> mf.run(50) # data from this will not be used in fitting a new MF model
+        >>> mf.run(50) # data from this WILL be used in fitting a new MF model
+        >>> mf.fit_mf()
+
+    """
+
+    def __init__(  # pylint: disable-all
+        self,
+        num_users=None,
+        num_items=None,
+        num_latent_factors=None,
+        user_representation=None,
+        item_representation=None,
+        actual_user_representation=None,
+        actual_item_representation=None,
+        probabilistic_recommendations=False,
+        seed=None,
+        verbose=False,
+        num_items_per_iter=10,
+        model_params=None,
+        **kwargs
+    ):
+        # check inputs
+        num_users, num_items, num_attributes = validate_user_item_inputs(
+            num_users,
+            num_items,
+            user_representation,
+            item_representation,
+            actual_user_representation,
+            actual_item_representation,
+            100,
```

Comment:
What are these values for?

Reply:
These values are default "fallback" values for number of users, number of items, and number of attributes. We do things this way (as opposed to having a simple default parameter in the method signature) because we have this very particular scheme of checking consistency between all inputs and then falling back to a default only if the argument cannot be inferred from other inputs. For example, `num_users` can sometimes be inferred from `user_representation`, but if we had `num_users=100` by default in the `__init__` signature, then it would be difficult to resolve conflicting values in the right "order."